README.md

# crossvalidation

Generic R functions for cross-validation 


### Installation

- __1st method__: from [R-universe](https://ropensci.org/r-universe/)

In R console:

```R
options(repos = c(
    techtonique = 'https://techtonique.r-universe.dev',
    CRAN = 'https://cloud.r-project.org'))
    
install.packages("crossvalidation")
```

- __2nd method__: from Github

In R console:

```R
devtools::install_github("Techtonique/crossvalidation")
```


### Demo

```R
# dataset
 set.seed(123)
 n <- 100 ; p <- 5
 X <- matrix(rnorm(n * p), n, p)
 y <- rnorm(n)
```

Linear model 

```R

# 'X' contains the explanatory variables
# 'y' is the response
# 'k' is the number of folds in k-fold cross-validation
# 'repeats' is the number of repeats of the k-fold cross-validation procedure

# linear model example -----

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3)

# linear model example, with validation set

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3, p = 0.8)
```

glmnet

```R
# glmnet example -----

# fit glmnet, with alpha = 1, lambda = 0.1

require(glmnet)
require(Matrix)

 crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
 fit_func = glmnet::glmnet, predict_func = predict.glmnet,
 packages = c("glmnet", "Matrix"), fit_params = list(alpha = 0.5, lambda = 0.1))

# fit glmnet, with alpha = 0, lambda = 0.01

 crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
 fit_func = glmnet::glmnet, predict_func = predict.glmnet,
 packages = c("glmnet", "Matrix"), fit_params = list(alpha = 0, lambda = 0.01))

 # fit glmnet, with alpha = 0, lambda = 0.01, with validation set

 crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 2, p = 0.8,
 fit_func = glmnet::glmnet, predict_func = predict.glmnet,
 packages = c("glmnet", "Matrix"), fit_params = list(alpha = 0, lambda = 0.01))
```

Random Forest

```R
# randomForest example -----

require(randomForest)

# fit randomForest with mtry = 2

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
fit_func = randomForest::randomForest, predict_func = predict,
packages = "randomForest", fit_params = list(mtry = 2))

# fit randomForest with mtry = 4

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
fit_func = randomForest::randomForest, predict_func = predict,
packages = "randomForest", fit_params = list(mtry = 4))

# fit randomForest with mtry = 4, with validation set

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 2, p = 0.8,
fit_func = randomForest::randomForest, predict_func = predict,
packages = "randomForest", fit_params = list(mtry = 4))
```

xgboost

```R
# xgboost example -----

require(xgboost)

# The response and covariates are named 'label' and 'data'
# So, we do this:

f_xgboost <- function(x, y, ...) xgboost::xgboost(data = x, label = y, ...)

# fit xgboost with nrounds = 5

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
  fit_func = f_xgboost, predict_func = predict,
   packages = "xgboost", fit_params = list(nrounds = 5,
   verbose = FALSE))

# fit xgboost with nrounds = 10

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
  fit_func = f_xgboost, predict_func = predict,
   packages = "xgboost", fit_params = list(nrounds = 10,
   verbose = FALSE))

# fit xgboost with nrounds = 10, with validation set

crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 2, p = 0.8,
  fit_func = f_xgboost, predict_func = predict,
   packages = "xgboost", fit_params = list(nrounds = 10,
   verbose = FALSE))
```

Theta method (time series)

```R
require(forecast)
res <- crossvalidation::crossval_ts(y=AirPassengers, initial_window = 10, fcast_func = thetaf)
print(colMeans(res))
```

### Contributing

Your contributions are welcome, and valuable. Please, make sure to __read__ the [Code of Conduct](CONTRIBUTING.md) first.

If you're not comfortable with Git/Version Control yet, please use [this form](https://forms.gle/nuKYGVc2HPxPUDEz7).

### License

[BSD 3-Clause](LICENSE) © Thierry Moudiki, 2019. 



CHANGES.md
# version 0.4.0 - November 14, 2021

- cross validation (in `crossval_ts`) on a percentage of the original 
dataset (training/testing dataset)
- evaluation on a __validation set__ (function `eval_ts`, must have the 
exact same `p` as `crossval_ts` to avoid overlapping slices)

# version 0.3.2

- Fix bug in `show_progress` (0% displayed when `show_progress==FALSE`)


tests/testthat.R
library(testthat)
library(crossvalidation)

test_check("crossvalidation")


tests/testthat/test_ml.R
context("Test on ML models")


library(testthat)
library(crossvalidation)

# dataset

 set.seed(123)
 n <- 1000 ; p <- 10
 X <- matrix(rnorm(n * p), n, p)
 y <- rnorm(n)

# linear model example -----

res1 <- crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3)


# randomForest example -----

require(randomForest)

# fit randomForest with mtry = 2

res2 <- crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
                   fit_func = randomForest::randomForest, predict_func = predict,
                   packages = "randomForest", fit_params = list(mtry = 2))

# fit randomForest with mtry = 4

res3 <- crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 3,
                   fit_func = randomForest::randomForest, predict_func = predict,
                   packages = "randomForest", fit_params = list(mtry = 4))

# fit randomForest with mtry = 4, with a validation set

res4 <- crossvalidation::crossval_ml(x = X, y = y, k = 5, repeats = 2, p = 0.8,
                   fit_func = randomForest::randomForest, predict_func = predict,
                   packages = "randomForest", fit_params = list(mtry = 4))


# tests -----

test_that("tests on mean and sd error", {
  expect_equal(round(res1$mean, 2), 1.01)
  expect_equal(round(res1$sd, 2), 0.03)
  expect_equal(round(res2$mean, 2), 1.02)
  expect_equal(round(res2$sd, 2), 0.03)
  expect_equal(round(res3$mean, 2), 1.02)
  expect_equal(round(res3$sd, 2),  0.03)
  expect_equal(round(res4$mean_training, 2), 1.03)
  expect_equal(round(res4$sd_training, 2), 0.04)
})



tests/testthat/test_time_series.R
context("Test on time series models")


library(testthat)
library(crossvalidation)
library(forecast)


# Example 1 -----

res1 <- crossval_ts(y=AirPassengers, initial_window = 10, fcast_func = thetaf)

res6 <- crossval_ts(y=AirPassengers, initial_window = 10, fcast_func = thetaf,
                    p = 0.8)


# Example 2 -----

fcast_func <- function (y, h, ...)
{
      forecast::forecast(forecast::auto.arima(y, ...),
      h=h, ...)
}

res2 <- crossval_ts(y=AirPassengers, initial_window = 10, fcast_func = fcast_func)



# Example 3 -----

fcast_func <- function (y, h, ...)
{
      forecast::forecast(forecast::ets(y, ...),
      h=h, ...)
}

res3 <- crossval_ts(y=AirPassengers, initial_window = 10, fcast_func = fcast_func)


# Example 4 -----

 set.seed(123)
 x <- ts(matrix(rnorm(50), nrow = 25))

 fcast_func <- function(y, h = 5, type_forecast=c("mean", "median"))
 {
  type_forecast <- match.arg(type_forecast)

  if (type_forecast == "mean")
  {
   means <- colMeans(y)
   return(list(mean = t(replicate(n = h, expr = means))))
  } else {
   medians <- apply(y, 2, median)
   return(list(mean = t(replicate(n = h, expr = medians))))
  }

 }

res4 <- crossvalidation::crossval_ts(y = x, fcast_func = fcast_func,
                                     fit_params = list(type_forecast = "median"))

res5 <- crossvalidation::crossval_ts(y = x, fcast_func = fcast_func,
                                     fit_params = list(type_forecast = "mean"))

res7 <- crossvalidation::crossval_ts(y = x, fcast_func = fcast_func,
                                     fit_params = list(type_forecast = "mean"),
                                     p = 0.8)

# Example 5 -----

set.seed(123)
res8 <- eval_ts(y=rnorm(25), initial_window = 10,
                horizon = 5, fcast_func = forecast::thetaf,
                show_progress = FALSE, q=0.2)

# tests -----

mean1 <- colMeans(res1)
mean2 <- colMeans(res2)
mean3 <- colMeans(res3)
mean4 <- colMeans(res4)
mean5 <- colMeans(res5)
mean8 <- colMeans(res8)

test_that("tests on mean RMSE and MAPE", {
  expect_equal(as.numeric(round(mean1["RMSE"], 2)), 51.43)
  expect_equal(as.numeric(round(mean1["MAPE"], 2)), 0.16 )
  expect_equal(as.numeric(round(mean2["RMSE"], 2)), 51.91)
  expect_equal(as.numeric(round(mean2["MAPE"], 2)), 0.15)
  expect_equal(as.numeric(round(mean3["RMSE"], 2)), 45.34)
  expect_equal(as.numeric(round(mean3["MAPE"], 2)), 0.14)
  expect_equal(as.numeric(round(mean4["RMSE"], 2)), 0.97)
  expect_equal(as.numeric(round(mean5["MAPE"], 2)), 1.43)
  expect_equal(res1[103, ], res6[103, ])
  expect_equal(res5[13, ], res7[13, ])
  expect_equal(as.numeric(round(mean8["RMSE"], 2)), 1.11)
  expect_equal(as.numeric(round(mean8["MAPE"], 2)), 1.35)
})



R/crossval.R
#' Generic cross-validation function
#'
#' Generic cross-validation
#'
#' @param x input covariates' matrix
#' @param y response variable; a vector
#' @param fit_func a function for fitting the model
#' @param predict_func a function for predicting values from the model
#' @param fit_params a list; additional (model-specific) parameters to be passed
#' to \code{fit_func}
#' @param k an integer; number of folds in k-fold cross validation
#' @param repeats an integer; number of repeats for the k-fold cross validation
#' @param p a double; proportion of data in the training/testing set, default is 1 and
#' must be > 0.5. If \code{p} < 1, a validation set error is calculated on the
#' remaining 1-\code{p} fraction data
#' @param seed random seed for reproducibility of results
#' @param eval_metric a function measuring the test errors; if not provided: RMSE for regression and
#' accuracy for classification
#' @param cl an integer; the number of clusters for parallel execution
#' @param errorhandling specifies how a task evalution error should be handled.
#' If value is "stop", then execution will be stopped if an error occurs. If value
#' is "remove", the result for that task will not be returned. If value is "pass",
#' then the error object generated by task evaluation will be included with the
#' rest of the results. The default value is "stop".
#' @param packages character vector of packages that the tasks depend on
#' @param verbose logical flag enabling verbose messages. This can be very useful for
#' trouble shooting
#' @param show_progress show evolution of the algorithm
#' @param ... additional parameters
#'
#' @return
#' @export
#'
#' @examples
#'
#'# dataset
#'
#' set.seed(123)
#' n <- 1000 ; p <- 10
#' X <- matrix(rnorm(n * p), n, p)
#' y <- rnorm(n)
#'
#'# linear model example -----
#'
#' crossval::crossval(x = X, y = y, k = 5, repeats = 3)
#'
#'
#'# randomForest example -----
#'
#'require(randomForest)
#'
#'# fit randomForest with mtry = 2
#'
#'crossval::crossval(x = X, y = y, k = 5, repeats = 3,
#'                   fit_func = randomForest::randomForest, predict_func = predict,
#'                   packages = "randomForest", fit_params = list(mtry = 2))
#'
#'# fit randomForest with mtry = 4
#'
#'crossval::crossval(x = X, y = y, k = 5, repeats = 3,
#'                   fit_func = randomForest::randomForest, predict_func = predict,
#'                   packages = "randomForest", fit_params = list(mtry = 4))
#'
#'# fit randomForest with mtry = 4, with a validation set
#'
#'crossval::crossval(x = X, y = y, k = 5, repeats = 2, p = 0.8,
#'                   fit_func = randomForest::randomForest, predict_func = predict,
#'                   packages = "randomForest", fit_params = list(mtry = 4))
#'
crossval <- function(x,
                     y,
                     fit_func = crossval::fit_lm,
                     predict_func = crossval::predict_lm,
                     fit_params = NULL,
                     # and hyperparameters
                     k = 5,
                     repeats = 3,
                     p = 1,
                     seed = 123,
                     eval_metric = NULL,
                     cl = NULL,
                     errorhandling = c('stop', 'remove', 'pass'),
                     packages = c("stats", "Rcpp"),
                     verbose = FALSE,
                     show_progress = TRUE,
                     ...) {
  n_y <- length(y)
  stopifnot(n_y == nrow(x))

  set.seed(seed)
  if (p == 1)
    # default
  {
    x <- as.matrix(x)
  } else {
    index_train <- sample.int(n_y, size = floor(p * n_y))
    x <- as.matrix(x[index_train,])
    y <- y[index_train]
    x_validation <- as.matrix(x[-index_train,])
    y_validation <- y[-index_train]
  }

  errorhandling <- match.arg(errorhandling)
  stopifnot(floor(k) == k || k > 10)
  stopifnot(p >= 0.5 && p <= 1)
  stopifnot(floor(repeats) == repeats)

  # evaluation metric for cv error
  if (is.null(eval_metric))
  {
    if (is.factor(y))
      # classification
    {
      eval_metric <- function (preds, actual)
      {
        res <- mean(preds == actual)
        names(res) <- "accuracy"
        return(res)
      }

    } else {
      # regression

      eval_metric <- function (preds, actual)
      {
        res <- sqrt(mean((preds - actual) ^ 2))
        names(res) <- "rmse"
        return(res)
      }

    }
  }

  set.seed(seed)
  list_folds <- lapply(1:repeats,
                       function (i)
                         crossval::create_folds(y = y, k = k))

  ptm <- proc.time()

  # parallel exec.
  if (!is.null(cl) && cl > 0)
  {
    cl_SOCK <- parallel::makeCluster(cl, type = "SOCK")
    doSNOW::registerDoSNOW(cl_SOCK)
    `%op1%` <-  foreach::`%dopar%`
    `%op2%` <-  foreach::`%do%`
    nb_iter <- k * repeats

    pb <- txtProgressBar(min = 0,
                         max = k,
                         style = 3)
    progress <- function(n)
      utils::setTxtProgressBar(pb, n)
    opts <- list(progress = progress)

    i <- NULL
    j <- NULL
    res <- foreach::foreach(
      i = 1:k,
      .packages = packages,
      .combine = rbind,
      .errorhandling = errorhandling,
      .options.snow = opts,
      .verbose = verbose,
      .export = c("create_folds")
    ) %op1% {
      foreach::foreach(
        j = 1:repeats,
        .packages = packages,
        .combine = cbind,
        .verbose = FALSE,
        .errorhandling = errorhandling,
        .export = c("fit_params")
      ) %op2% {
        train_index <- -list_folds[[j]][[i]]
        test_index <-
          -train_index

        # fit
        set.seed(seed) # in case the algo is randomized
        fit_func_train <-
          function(x, y, ...)
            fit_func(x = x[train_index,],
                     y = y[train_index],
                     ...)

        fit_obj <-
          do.call(what = fit_func_train,
                  args = c(list(x = x[train_index,],
                                y = y[train_index]),
                           fit_params))

        # predict
        preds <-
          try(predict_func(fit_obj, newdata = x[test_index,]),
              silent = TRUE)
        if (class(preds) == "try-error")
        {
          preds <- try(predict_func(fit_obj, newx = x[test_index,]),
                       silent = TRUE)
          if (class(preds) == "try-error")
          {
            preds <- rep(NA, length(test_index))
          }
        }

        # measure the error
        error_measure <-
          eval_metric(preds, y[test_index])

        if (show_progress)
        {
          setTxtProgressBar(pb, i * j)
        }

        if (p == 1) {
          error_measure

        } else {
          # there is a validation set

          # predict on validation set
          preds_validation <-
            try(predict_func(fit_obj,
                             newdata = x_validation),
                silent = TRUE)

          if (class(preds_validation) == "try-error")
          {
            preds_validation <- try(predict_func(fit_obj,
                                                 newx = x_validation),
                                    silent = TRUE)

            if (class(preds_validation) == "try-error")
            {
              preds_validation <- rep(NA, length(y_validation))
            }
          }

          # measure the validation error
          c(error_measure,
            eval_metric(preds_validation, y_validation))
        }

      }

    }
    close(pb)
    snow::stopCluster(cl_SOCK)

  }  else {
    # sequential exec.

    `%op%` <-  foreach::`%do%`

    pb <- txtProgressBar(min = 0,
                         max = k,
                         style = 3)
    progress <- function(n)
      utils::setTxtProgressBar(pb, n)

    i <- NULL
    res <- foreach::foreach(
      i = 1:k,
      .packages = packages,
      .combine = rbind,
      .errorhandling = errorhandling,
      .verbose = verbose,
      .export = c("create_folds")
    ) %op% {
      if (show_progress)
      {
        setTxtProgressBar(pb, i)
      }

      temp <-
        foreach::foreach(
          j = 1:repeats,
          .packages = packages,
          .combine = cbind,
          .verbose = FALSE,
          .errorhandling = errorhandling,
          .export = c("fit_params")
        ) %op% {
          train_index <- -list_folds[[j]][[i]]
          test_index <-
            -train_index

          # fit
          set.seed(seed) # in case the algo is randomized
          fit_func_train <-
            function(x, y, ...)
              fit_func(x = x[train_index,],
                       y = y[train_index],
                       ...)

          fit_obj <-
            do.call(what = fit_func_train,
                    args = c(list(x = x[train_index,],
                                  y = y[train_index]),
                             fit_params))

          # predict
          preds <-
            try(predict_func(fit_obj, newdata = x[test_index,]),
                silent = TRUE)
          if (class(preds) == "try-error")
          {
            preds <- try(predict_func(fit_obj, newx = x[test_index,]),
                         silent = TRUE)
            if (class(preds) == "try-error")
            {
              preds <- rep(NA, length(test_index))
            }
          }

          # measure the error
          error_measure <-
            eval_metric(preds, y[test_index])

          if (p == 1) {
            error_measure

          } else {
            # there is a validation set

            # predict on validation set
            preds_validation <-
              try(predict_func(fit_obj,
                               newdata = x_validation),
                  silent = TRUE)

            if (class(preds_validation) == "try-error")
            {
              preds_validation <- try(predict_func(fit_obj,
                                                   newx = x_validation),
                                      silent = TRUE)

              if (class(preds_validation) == "try-error")
              {
                preds_validation <- rep(NA, length(y_validation))
              }
            }

            # measure the validation error
            c(error_measure,
              eval_metric(preds_validation, y_validation))
          }

        }

    }

  }

  if (show_progress)
  {
    cat("\n")
    print(proc.time() - ptm)
    cat("\n")
  }

  if (p == 1)
  {
    colnames(res) <- paste0("repeat_", 1:ncol(res))
    rownames(res) <- paste0("fold_", 1:nrow(res))

    return(list(
      folds = res,
      mean = mean(res, na.rm = TRUE),
      sd = sd(res, na.rm = TRUE),
      median = median(res, na.rm = TRUE)
    ))
  } else {
    if (repeats > 1)
    {
      colnames(res) <- paste0("repeat_", 1:ncol(res))
      rownames(res) <-
        paste0(rep(c(
          "fold_training_", "fold_validation_"
        ), k),
        rep(1:k, each = 2))
    } else {
      res <- as.numeric(res)
      names(res) <-
        paste0(rep(c(
          "fold_training_", "fold_validation_"
        ), k),
        rep(1:k, each = 2))
    }

    n_folds <- nrow(res)
    train_test_df <- res[seq(1, n_folds, by = 2), ]
    validation_df <- res[seq(2, n_folds, by = 2), ]

    return(
      list(
        folds = res,
        mean_training = mean(train_test_df),
        mean_validation = mean(validation_df),
        sd_training = sd(train_test_df),
        sd_validation = sd(validation_df),
        median_training = median(train_test_df),
        median_validation = median(validation_df)
      )
    )
  }

}
compiler::cmpfun(crossval)


R/create_folds.R
# inspired from caret::createFolds
create_folds <- function(y, k = 10)
{
  if (is.numeric(y)) {
    cuts <- floor(length(y)/k)
    if (cuts < 2)
      cuts <- 2
    if (cuts > 5)
      cuts <- 5
    breaks <- unique(quantile(y, probs = seq(0, 1, length = cuts)))
    y <- cut(y, breaks, include.lowest = TRUE)
  }
  if (k < length(y)) {
    y <- factor(as.character(y))
    numInClass <- table(y)
    foldVector <- vector(mode = "integer", length(y))
    for (i in 1:length(numInClass)) {
      min_reps <- numInClass[i]%/%k
      if (min_reps > 0) {
        spares <- numInClass[i]%%k
        seqVector <- rep(1:k, min_reps)
        if (spares > 0)
          seqVector <- c(seqVector, sample(1:k, spares))
        foldVector[which(y == names(numInClass)[i])] <- sample(seqVector)
      }
      else {
        foldVector[which(y == names(numInClass)[i])] <- sample(1:k,
                                                               size = numInClass[i])
      }
    }
  }
  else foldVector <- seq(along = y)

  out <- split(seq(along = y), foldVector)
  names(out) <- paste("Fold", gsub(" ", "0", format(seq(along = out))),
                      sep = "")


  return(out)

}
create_folds <- compiler::cmpfun(create_folds)


# borrowed from caret::createTimeSlices

create_time_slices <- function(y, initial_window, horizon = 1,
                               fixed_window = TRUE, skip = 0)
{
  if(!is.null(ncol(y)))
  {
    n_y <- dim(y)[1]
  } else {
    n_y <- length(y)
  }

  stops <- seq(initial_window, (n_y - horizon), by = skip + 1)

  if (fixed_window) {
    starts <- stops - initial_window + 1
  }
  else {
    starts <- rep(1, length(stops))
  }
  train <- mapply(seq, starts, stops, SIMPLIFY = FALSE)
  test <- mapply(seq, stops + 1, stops + horizon, SIMPLIFY = FALSE)
  nums <- gsub(" ", "0", format(stops))
  names(train) <- paste("training", nums, sep = "")
  names(test) <- paste("testing", nums, sep = "")
  out <- list(train = train, test = test)
  out
}
create_time_slices  <- compiler::cmpfun(create_time_slices)


R/eval_ts.R
#' Rolling origin evaluation on validation set (time series)
#'
#' Rolling origin evaluation on validation set (time series)
#'
#' @param y response time series; a vector or a matrix
#' @param x input covariates' matrix (optional) for ML models
#' @param fit_func a function for fitting the model (if validation of ML model)
#' @param predict_func a function for predicting values from the model (if validation of ML model)
#' @param fcast_func time series forecasting function (e.g forecast::thetaf)
#' @param fit_params a list; additional (model-specific) parameters to be passed
#' to \code{fit_func}
#' @param q a float; percentage of original data in the validation test.
#' @param initial_window an integer; the initial number of consecutive values in each training set sample
#' @param horizon an integer; the number of consecutive values in test set sample
#' @param fixed_window a boolean; if FALSE, all training samples start at 1
#' @param level a numeric vector; confidence levels for prediction intervals.
#' @param seed random seed for reproducibility of results
#' @param eval_metric a function measuring the test errors; if not provided: RMSE for regression and
#' accuracy for classification
#' @param cl an integer; the number of clusters for parallel execution
#' @param errorhandling specifies how a task evalution error should be handled.
#' If value is "stop", then execution will be stopped if an error occurs. If value
#' is "remove", the result for that task will not be returned. If value is "pass",
#' then the error object generated by task evaluation will be included with the
#' rest of the results. The default value is "stop".
#' @param packages character vector of packages that the tasks depend on
#' @param verbose logical flag enabling verbose messages. This can be very useful for
#' troubleshooting.
#' @param show_progress show evolution of the algorithm
#' @param ... additional parameters
#'
#' @return
#' @export
#'
#' @examples
#'
#'
#' require(forecast)
#' data("AirPassengers")
#'
#' # Example 1 -----
#'
#' res <- eval_ts(y=AirPassengers, initial_window = 10,
#' horizon = 3, fcast_func = forecast::thetaf)
#' print(colMeans(res))
#'
#'
#' # Example 2 -----
#'
#' \dontrun{
#' fcast_func <- function (y, h, ...)
#' {
#'       forecast::forecast(forecast::auto.arima(y, ...),
#'       h=h, ...)
#' }
#'
#' res <- eval_ts(y=AirPassengers, initial_window = 10, horizon = 3,
#' fcast_func = fcast_func)
#' print(colMeans(res))
#' }
#'
#'
#' # Example 3 -----
#'
#' fcast_func <- function (y, h, ...)
#' {
#'       forecast::forecast(forecast::ets(y, ...),
#'       h=h, ...)
#' }
#'
#' res <- eval_ts(y=AirPassengers,
#' initial_window = 10, horizon = 3, fcast_func = fcast_func)
#' print(colMeans(res))
#'
#'
#' # Example 4 -----
#'
#' xreg <- cbind(1, 1:length(AirPassengers))
#' res <- eval_ts(y=AirPassengers, x=xreg, fit_func = crossvalidation::fit_lm,
#' predict_func = crossvalidation::predict_lm,
#' initial_window = 10,
#' horizon = 3,
#' fixed_window = TRUE)
#' print(colMeans(res))
#'
#'
#' # Example 5 -----
#'
#' res <- eval_ts(y=AirPassengers, fcast_func = forecast::thetaf,
#' initial_window = 10,
#' horizon = 3,
#' fixed_window = TRUE)
#' print(colMeans(res))
#'
#'
#'#' # Example 6 -----
#'
#' xreg <- cbind(1, 1:length(AirPassengers))
#' res <- eval_ts(y=AirPassengers, x=xreg, fit_func = crossvalidation::fit_lm,
#' predict_func = crossvalidation::predict_lm,
#' initial_window = 10,
#' horizon = 3,
#' fixed_window = TRUE)
#' print(colMeans(res))
#'
#'
#' # Example 7 -----
#'
#' x <- ts(matrix(rnorm(50), nrow = 25))
#'
#' fcast_func <- function(y, h = 5, type_forecast=c("mean", "median"))
#' {
#'  type_forecast <- match.arg(type_forecast)
#'
#'  if (type_forecast == "mean")
#'  {
#'   means <- colMeans(y)
#'   return(list(mean = t(replicate(n = h, expr = means))))
#'  } else {
#'   medians <- apply(y, 2, median)
#'   return(list(mean = t(replicate(n = h, expr = medians))))
#'  }
#'
#' }
#'
#' print(fcast_func(x))
#'
#' res <- crossvalidation::eval_ts(y = x, fcast_func = fcast_func, fit_params = list(type_forecast = "median"))
#' colMeans(res)
#'
#' res <- crossvalidation::eval_ts(y = x, fcast_func = fcast_func, fit_params = list(type_forecast = "mean"))
#' colMeans(res)
#'
#' # Example 8 -----
#'
#' eval_metric <- function(predicted, observed)
#' {
#'   error <- observed - predicted
#'
#'   res <- apply(error, 2, function(x) sqrt(mean(x ^ 2, na.rm = FALSE)))
#'
#'   return(res)
#' }
#'
#' res <- crossvalidation::eval_ts(y = x, fcast_func = fcast_func, fit_params = list(type_forecast = "mean"),
#' eval_metric = eval_metric)
#'
#' colMeans(res)
#'
eval_ts <- function(y,
                    x = NULL,
                    fit_func = crossvalidation::fit_lm,
                    predict_func = crossvalidation::predict_lm,
                    fcast_func = NULL,
                    fit_params = NULL,
                    q = 0.2,
                    # parameters of funcs
                    initial_window = 5,
                    horizon = 3,
                    fixed_window = TRUE,
                    level = c(80, 95),
                    seed = 123,
                    eval_metric = NULL,
                    cl = NULL,
                    errorhandling = c('stop', 'remove', 'pass'),
                    packages = c("stats", "Rcpp"),
                    verbose = FALSE,
                    show_progress = TRUE,
                    ...) {

  p <- 1 - q

  stopifnot(p < 1)

  # define training set (to be excluded from  slices)
  if(!is.null(ncol(y))) # multivariate input y
  {
    if (is.ts(y))
    {
      y_train <- ts(y[1:floor(p*nrow(y)), ],
              start = start(y),
              frequency = frequency(y))
    } else {
      y_train <- ts(y[1:floor(p*nrow(y)), ])
    }

    time_slices_total <-
      crossvalidation::create_time_slices(
        y[, 1],
        initial_window = initial_window,
        horizon = horizon,
        fixed_window = fixed_window
      )

    time_slices_train <-
      crossvalidation::create_time_slices(
        y_train[, 1],
        initial_window = initial_window,
        horizon = horizon,
        fixed_window = fixed_window
      )

  } else { # univariate input y

    if (is.ts(y))
    {
      y_train <- ts(y[1:floor(p*length(y))],
              start = start(y),
              frequency = frequency(y))
    } else {
      y_train <- ts(y[1:floor(p*length(y))])
    }

    time_slices_total <-
      crossvalidation::create_time_slices(
        y,
        initial_window = initial_window,
        horizon = horizon,
        fixed_window = fixed_window
      )

    time_slices_train <-
      crossvalidation::create_time_slices(
        y_train,
        initial_window = initial_window,
        horizon = horizon,
        fixed_window = fixed_window
      )
  }

  # cat("time_slices_total", "\n")
  # print(time_slices_total$train)
  # cat("time_slices_train", "\n")
  # print(time_slices_train$train)
  # cat("\n")

  time_slices <- list()

  time_slices$train <- time_slices_total$train[!(time_slices_total$train %in% time_slices_train$train)]

  time_slices$test <- time_slices_total$test[!(time_slices_total$test %in% time_slices_train$test)]

  # cat("time_slices", "\n")
  # print(time_slices$train)
  # cat("\n")

  n_slices <- length(time_slices$train)

  if (!is.null(x)) # regression, ML model
  {
    n_x <- dim(x)[1]
    p_x <- dim(x)[2]
   # stopifnot(n_x == n_y)
  }

  # performance metrics
  if (is.null(eval_metric))
  {
    eval_metric <- function(predicted, observed)
    {
      error <- observed - predicted
      pe <- predicted / observed - 1

      res <- c(
        mean(error, na.rm = FALSE),
        sqrt(mean(error ^ 2, na.rm = FALSE)),
        mean(abs(error), na.rm = FALSE),
        mean(pe, na.rm = FALSE),
        mean(abs(pe), na.rm = FALSE)
      )

      names(res) <- c("ME", "RMSE", "MAE", "MPE", "MAPE")

      return(res)
    }
    eval_metric <- compiler::cmpfun(eval_metric)
  }

  # progress bars
  if (!is.null(cl)) {
    cl_SOCK <- parallel::makeCluster(cl, type = "SOCK")
    doSNOW::registerDoSNOW(cl_SOCK)
    `%op%` <- foreach::`%dopar%`
  } else {
    `%op%` <- foreach::`%do%`
  }

  if (show_progress)
  {
    pb <- txtProgressBar(min = 0,
                         max = n_slices,
                         style = 3)
    progress <- function(n) {utils::setTxtProgressBar(pb, n)}
    opts <- list(progress = progress)
  } else {
    opts <- NULL
  }


  if (!is.null(fcast_func)) { # if fcast_func is not NULL, ts models are used

    # 1 - interface for forecasting functions --------------------------------------------------

    i <- NULL
    res <- foreach::foreach(
      i = 1:n_slices,
      .packages = packages,
      .combine = rbind,
      .errorhandling = errorhandling,
      .options.snow = opts,
      .verbose = verbose
    ) %op% {

      train_index <- time_slices$train[[i]]
      test_index <- time_slices$test[[i]]

      # cat("train_index", "\n")
      # print(train_index)
      # cat("\n")
      #
      # cat("test_index", "\n")
      # print(test_index)
      # cat("\n")


      # 1 - 1 interface for forecasting functions: univariate --------------------------------------------------

      if (is.null(ncol(y))) # univariate time series case
      {
        preds <- try(do.call(what = fcast_func,
                             args = c(list(y = y[train_index],
                                           h = horizon), fit_params))$mean, silent = FALSE)

        if (class(preds)[1] == "try-error")
        {
          preds <- rep(NA, horizon)
        }

        # measure the error
        error_measure <-
          eval_metric(preds, y[test_index]) # univariate

      } else { #multivariate time series case

        # 1 - 2 interface for forecasting functions: multivariate --------------------------------------------------

        preds <- try(do.call(
          what = fcast_func,
          args = c(list(y = y[train_index, ],
                        h = horizon), fit_params)
        )$mean, silent = FALSE)

        if (class(preds)[1] == "try-error" | is.null(preds))
        {
          preds <- rep(NA, horizon)
        }

        # measure the error
        error_measure <-
          eval_metric(preds, y[test_index, ])

      }

      if (show_progress)
      {
        setTxtProgressBar(pb, i)
      }

      error_measure

    }

    if (show_progress)
    {
      close(pb)
    }

    if (!is.null(cl))
    {
      snow::stopCluster(cl_SOCK)
    }

  } else { # if fcast_func is NULL, ML models are used

    stopifnot(!is.null(fit_func))
    stopifnot(!is.null(predict_func))

    # 2 - interface for ml functions --------------------------------------------------

    i <- NULL
    res <- foreach::foreach(
      i = 1:n_slices,
      .packages = packages,
      .combine = rbind,
      .errorhandling = errorhandling,
      .options.snow = opts,
      .verbose = verbose
    ) %op% {
      # predict
      train_index <-
        time_slices$train[[i]]
      test_index <- time_slices$test[[i]]


      if (is.null(ncol(y)))
      {
        # 2 - 1 interface for ml function: univariate --------------------------------------------------
        fit_obj <-
          do.call(what = fit_func,
                  args = c(list(x = x[train_index,],
                                y = y[train_index]),
                           fit_params))

        # predict
        preds <-
          try(predict_func(fit_obj, newdata = x[test_index,]),
              silent = TRUE)

        if (class(preds)[1] == "try-error")
        {
          preds <- try(predict_func(fit_obj, newx = x[test_index,]),
                       silent = FALSE)
          if (class(preds) == "try-error")
          {
            preds <- rep(NA, length(test_index))
          }
        }

        # measure the error
        error_measure <-
          eval_metric(preds, y[test_index])

      } else {

        # 2 - 2 interface for ml function: multivariate (ko so far) --------------------------------------------------
        stop("Not implemented")

      }

      if (show_progress)
      {
        setTxtProgressBar(pb, i)
      }

      error_measure
    }

    if (show_progress)
    {
      close(pb)
    }

    if (!is.null(cl))
    {
      snow::stopCluster(cl_SOCK)
    }
  }



  return(res)

}
compiler::cmpfun(eval_ts)
